### PR TITLE
feat: reverse proxy for text/event-stream

### PR DESCRIPTION
**Pull request by dspo**
#### What this PR does / why we need it:
feat: reverse proxy for text/event-stream

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](paste your link here)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English  | feat: reverse proxy for text/event-stream |
| 🇨🇳 中文    | 反向代理框架支持 text/event-stream 流式返回 |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: cmd/ai-proxy/conf/routes.yml [hunk 1]
```diff
@@ -134,15 +121,10 @@ routes:
...
-          to: provider
-          provider: azure
-          instanceId: default
-          scheme: https
-          host: ""
-          path: ""
-  - path: /openai/deployments/{deployment-id}/embeddings
+
+  - path: /openai/deployments/{deploymentId}/embeddings
+    router:
```
Comment (sfwn) [inline cmd/ai-proxy/conf/routes.yml:126]: why not use `/v1/embeddings`

File: internal/apps/ai-proxy/filters/audit/filter.go [hunk 2]
```diff
@@ -54,100 +54,126 @@ const (
...
-	return &Audit{Audit: new(models.AIProxyFilterAudit)}, nil
+func New(_ json.RawMessage) (reverseproxy.Filter, error) {
+	return &Audit{Audit: new(models.AIProxyFilterAudit), DefaultResponseFilter: reverseproxy.NewDefaultResponseFilter()}, nil
 }
 
-func (f *Audit) OnHttpRequestInfor(ctx context.Context, g filter.HttpInfor) (filter.Signal, error) {
-	var l = ctx.Value(filter.LoggerCtxKey{}).(logs.Logger).Sub("Audit").Sub("OnHttpRequestInfor")
-	for _, set := range []func(context.Context, http.Header, *bytes.Buffer) error{
+func (f *Audit) OnRequest(ctx context.Context, w http.ResponseWriter, infor reverseproxy.HttpInfor) (signal reverseproxy.Signal, err error) {
+	var l = ctx.Value(reverseproxy.LoggerCtxKey{}).(logs.Logger).Sub("Audit").Sub("OnHttpRequestInfor")
```
Comment (sfwn) [inline internal/apps/ai-proxy/filters/audit/filter.go:76]: mismatched log sub.
you should provider a method to get logger and inject subs using reflect.

File: pkg/reverseproxy/reverseproxy.go [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+	removeConnectionHeaders(outreq.Header)
+
+	// Remove hop-by-hop headers to the backend. Especially
+	// important is "Connection" because we want a persistent
+	// connection, regardless of what the client sent to us.
+	for _, h := range hopHeaders {
+		outreq.Header.Del(h)
+	}
+
+	// Issue 21096: tell backend applications that care about trailer support
```
Comment (sfwn) [inline pkg/reverseproxy/reverseproxy.go:338]: whose issue?

File: pkg/reverseproxy/reverseproxy.go [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+// If an X-Forwarded-For header already exists, the client IP is
+// appended to the existing values. As a special case, if the header
+// exists in the Request.Header map but has a nil value (such as when
+// set by the Director func), the X-Forwarded-For header is
+// not modified.
+//
+// To prevent IP spoofing, be sure to delete any pre-existing
+// X-Forwarded-For header coming from the client or
+// an untrusted proxy.
+type ReverseProxy struct {
```
Comment (sfwn) [inline pkg/reverseproxy/reverseproxy.go:55]: Why not just embed the official reverse proxy?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: cmd/ai-proxy/conf/routes.yml [hunk 5]
```diff
@@ -10,14 +19,13 @@ routes:
...
             - SetOpenAIOrganizationIfNotSpecified
-      - name: reverse-proxy
-        config:
-          to: provider # enum: "provider", "url"
-          provider: openai
-          instanceId: default
-          scheme: https
+
   - path: /v1/models/{model}
+    pathMatcher: ""   # 不填写时, 默认为 `^/v1/models/([^/.]+)$`
```
[the remark this answers is not in the document]
Comment (dspo) [reply cmd/ai-proxy/conf/routes.yml:24]: 因为 path 里有占位符 `{model}` 所以会转换成正则表达式 `^/v1/models/(?P<model>[^/.]+)$` 这样去匹配

File: cmd/ai-proxy/conf/routes.yml (hunk 1 above)
Comment (dspo) [reply cmd/ai-proxy/conf/routes.yml:126]: openai 没有 `/v1/embeddings` 这样的接口, `/openai/deployments/{deploymentId}/embeddings` 接口是 azure 特有的

Comment (dspo) [reply cmd/ai-proxy/conf/routes.yml:126]: 用到 `/v1/embeddings` 这个接口的时候再加

File: internal/apps/ai-proxy/filters/audit/filter.go (hunk 2 above)
Comment (dspo) [reply internal/apps/ai-proxy/filters/audit/filter.go:76]: 记个任务, 优化一下

**Review by sfwn — APPROVED** [review]

File: internal/apps/ai-proxy/metrics/collector.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+			value.Count++
+		} else {
+			m[key] = &lvsDistincter{LVs: lv, Count: 1}
+		}
+	}
+	for _, value := range m {
+		ch <- prometheus.MustNewConstMetric(c.desc, prometheus.CounterValue, value.Count, value.LVs.Values()...)
+	}
+}
+
```
Comment (dspo) [inline internal/apps/ai-proxy/metrics/collector.go:78]: @sfwn 这会导致指标变成两倍吗 ?

**Pull request merged**